components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java

```diff
@@ -6738,4 +6812,122 @@ private void rollbackAddApplicationTransaction(Connection connection, ServicePro
             throw new IdentityApplicationManagementException("Error while rolling back the transaction.", e);
         }
     }
+
+    /**
+     * Retrieve the discoverable groups for the application.
+     *
+     * @param applicationId Application ID.
+     * @param connection    Database connection.
+     * @param tenantDomain  Tenant domain of the application.
+     * @return Discoverable groups list.
+     * @throws IdentityApplicationManagementException If an error occurred while retrieving discoverable groups.
+     */
+    private DiscoverableGroup[] getDiscoverableGroups(int applicationId, Connection connection, String tenantDomain)
+            throws IdentityApplicationManagementException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Retrieving discoverable groups for application with ID: " + applicationId);
+        }
+
+        Map<String, List<GroupBasicInfo>> groupInfoMap = new HashMap<>();
+        try (PreparedStatement statement = connection.prepareStatement(
+                ApplicationMgtDBQueries.GET_GROUP_ASSOCIATIONS_BY_APP_ID)) {
+            statement.setInt(1, applicationId);
+            try (ResultSet resultSet = statement.executeQuery()) {
+                AbstractUserStoreManager userStoreManager =
+                        ApplicationMgtUtil.getUserStoreManager(tenantDomain);
+                while (resultSet.next()) {
+                    String groupID = resultSet.getString(1);
+                    String domainName = resultSet.getString(2);
+                    GroupBasicInfo groupBasicInfo = new GroupBasicInfo();
+                    groupBasicInfo.setId(groupID);
+                    try {
+                        String groupName = userStoreManager.getGroupNameByGroupId(groupID);
+                        groupBasicInfo.setName(UserCoreUtil.removeDomainFromName(groupName));
+                        if (groupInfoMap.containsKey(domainName)) {
+                            groupInfoMap.get(domainName).add(groupBasicInfo);
+                        } else {
+                            List<GroupBasicInfo> groupBasicInfoList = new ArrayList<>();
+                            groupBasicInfoList.add(groupBasicInfo);
+                            groupInfoMap.put(domainName, groupBasicInfoList);
+                        }
+                    } catch (UserStoreException e) {
+                        log.warn("Error while retrieving group name for group ID: " + groupID, e);
+                    }
+                }
+            }
+        } catch (SQLException e) {
+            throw new IdentityApplicationManagementException(
+                    "Error while retrieving discoverable groups for the application", e);
+        }
+        List<DiscoverableGroup> discoverableGroups = new ArrayList<>();
```

Comment:
refactor the hash map

Reply:
https://github.com/wso2/carbon-identity-framework/pull/6283/commits/25cd6027c99764766e277a7a1d9987e830cc3b4e